File: ios/Sources/PoltioSDK/PoltioLogger.swift
import Foundation

/// Represents the verbosity level of internal SDK logging.
@objc public enum PoltioLogLevel: Int, Comparable {
    /// Disables all SDK console logging.
    case none = 0
    /// Logs only critical errors and failures.
    case error = 1
    /// Logs warnings and critical errors.
    case warning = 2
    /// Logs informational events, state changes, and errors (default).
    case info = 3
    /// Logs detailed debug traces, network payloads, and internal transitions.
    case debug = 4

    public static func < (lhs: PoltioLogLevel, rhs: PoltioLogLevel) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

/// Internal thread-safe logger for Poltio SDK.
public enum PoltioLogger {
    private static let lock = NSLock()
    private static var _logLevel: PoltioLogLevel = .info

    /// The current active log level for the SDK.
    public static var logLevel: PoltioLogLevel {
        get {
            lock.lock()
            defer { lock.unlock() }
            return _logLevel
        }
        set {
            lock.lock()
            _logLevel = newValue
            lock.unlock()
        }
    }

    /// Logs a debug message if the active log level is `.debug`.
    public static func debug(_ message: @autoclosure () -> String) {
        log(.debug, message())
    }

    /// Logs an informational message if the active log level is `.info` or higher.
    public static func info(_ message: @autoclosure () -> String) {
        log(.info, message())
    }

    /// Logs a warning message if the active log level is `.warning` or higher.
    public static func warning(_ message: @autoclosure () -> String) {
        log(.warning, message())
    }

    /// Logs an error message if the active log level is `.error` or higher.
    public static func error(_ message: @autoclosure () -> String) {
        log(.error, message())
    }

    /// Logs a message at the specified level if allowed by the active log level.
    public static func log(_ level: PoltioLogLevel, _ message: @autoclosure () -> String) {
        guard level != .none, level <= logLevel else { return }
        print("[PoltioSDK] \(message())")
    }
}

File: android/poltio-sdk/src/main/java/com/poltio/sdk/PoltioLogger.kt
package com.poltio.sdk

import android.util.Log

/** Verbosity level of internal SDK logging. */
enum class PoltioLogLevel(val rawValue: Int) : Comparable<PoltioLogLevel> {
    /** Disables all SDK logging. */
    NONE(0),

    /** Logs only critical errors and failures. */
    ERROR(1),

    /** Logs warnings and critical errors. */
    WARNING(2),

    /** Logs informational events, state changes, and errors (default). */
    INFO(3),

    /** Logs detailed debug traces, network payloads, and internal transitions. */
    DEBUG(4),
}

/**
 * Internal thread-safe logger for the Poltio SDK. This is the single sanctioned sink for SDK
 * output — no other SDK source file should call `Log.*`/`println` directly.
 */
object PoltioLogger {
    private const val TAG = "PoltioSDK"

    @Volatile
    var logLevel: PoltioLogLevel = PoltioLogLevel.INFO

    /** Logs a debug message if the active log level is `.DEBUG`. Message is lazily built. */
    fun debug(message: () -> String) = log(PoltioLogLevel.DEBUG, message)

    /** Logs an informational message if the active log level is `.INFO` or higher. */
    fun info(message: () -> String) = log(PoltioLogLevel.INFO, message)

    /** Logs a warning message if the active log level is `.WARNING` or higher. */
    fun warning(message: () -> String) = log(PoltioLogLevel.WARNING, message)

    /** Logs an error message if the active log level is `.ERROR` or higher. */
    fun error(message: () -> String) = log(PoltioLogLevel.ERROR, message)

    /** Logs a message at the specified level if allowed by the active log level. */
    fun log(level: PoltioLogLevel, message: () -> String) {
        if (level == PoltioLogLevel.NONE || level > logLevel) return
        val text = message()
        when (level) {
            PoltioLogLevel.DEBUG -> Log.d(TAG, text)
            PoltioLogLevel.INFO -> Log.i(TAG, text)
            PoltioLogLevel.WARNING -> Log.w(TAG, text)
            PoltioLogLevel.ERROR -> Log.e(TAG, text)
            PoltioLogLevel.NONE -> Unit
        }
    }
}
